File: frontend/src/services/orderService.ts
import { Order, OrderWithDetails, OrderRequest, ApiResponse } from '../types'
import api from './api'

export const orderService = {
  async getOrders(): Promise<OrderWithDetails[]> {
    const response = await api.get<ApiResponse<OrderWithDetails[]>>('/orders')
    return response.data.data || []
  },

  async getOrder(id: number): Promise<OrderWithDetails> {
    const response = await api.get<ApiResponse<OrderWithDetails>>(`/orders/${id}`)
    if (!response.data.data) {
      throw new Error('Order not found')
    }
    return response.data.data
  },

  async createOrder(data: OrderRequest): Promise<Order> {
    const response = await api.post<ApiResponse<Order>>('/orders', data)
    if (!response.data.data) {
      throw new Error('Failed to create order')
    }
    return response.data.data
  },

  async updateOrder(id: number, data: OrderRequest): Promise<Order> {
    const response = await api.put<ApiResponse<Order>>(`/orders/${id}`, data)
    if (!response.data.data) {
      throw new Error('Failed to update order')
    }
    return response.data.data
  },

  async deleteOrder(id: number): Promise<void> {
    await api.delete(`/orders/${id}`)
  },

  async getOrdersByCustomer(customerId: number): Promise<OrderWithDetails[]> {
    const response = await api.get<ApiResponse<OrderWithDetails[]>>(`/orders/customer/${customerId}`)
    return response.data.data || []
  },

  async updateOrderStatus(id: number, status: string): Promise<void> {
    await api.patch(`/orders/${id}/status`, { status })
  }
}

File: backend/internal/services/order_service.go
package services

import (
	"wms-backend/internal/models"
	"wms-backend/internal/repositories"
)

type OrderService interface {
	GetAll() ([]models.OrderWithDetails, error)
	GetByID(id int) (*models.OrderWithDetails, error)
	Create(data *models.OrderRequest) (*models.Order, error)
	Update(id int, data *models.OrderRequest) (*models.Order, error)
	Delete(id int) error
	GetByCustomerID(customerID int) ([]models.OrderWithDetails, error)
	UpdateStatus(id int, status string) error
}

type orderService struct {
	orderRepo repositories.OrderRepository
}

func NewOrderService(orderRepo repositories.OrderRepository) OrderService {
	return &orderService{
		orderRepo: orderRepo,
	}
}

func (s *orderService) GetAll() ([]models.OrderWithDetails, error) {
	return s.orderRepo.GetAll()
}

func (s *orderService) GetByID(id int) (*models.OrderWithDetails, error) {
	return s.orderRepo.GetByID(id)
}

func (s *orderService) Create(data *models.OrderRequest) (*models.Order, error) {
	return s.orderRepo.Create(data)
}

func (s *orderService) Update(id int, data *models.OrderRequest) (*models.Order, error) {
	return s.orderRepo.Update(id, data)
}

func (s *orderService) Delete(id int) error {
	return s.orderRepo.Delete(id)
}

func (s *orderService) GetByCustomerID(customerID int) ([]models.OrderWithDetails, error) {
	return s.orderRepo.GetByCustomerID(customerID)
}

func (s *orderService) UpdateStatus(id int, status string) error {
	return s.orderRepo.UpdateStatus(id, status)
}
